Add more execution receipt checks

liuchengxu opened the pull request:
This PR is a follow-up of #427 that adds some more necessary checks for the receipt. The main points are:

- Ensure the block that the receipt was created from is part of the history, i.e., the header of the block specified by the receipt primary number exists.
- Ensure the receipt in the tx pool is neither stale nor too far in the future
    - `receipt_primary_number ∈ (best_execution_chain_number, best_execution_chain_number + max_drift]`
    - Add proper `provides` and `requires` tags to the unsigned extrinsic, similar to the nonce strategy to a large degree.
- Ensure the receipt primary number is exactly `best_execution_chain_number + 1` before actually dispatching the extrinsic.
- Revert the execution chain if the fraud proof is valid so that the new correct ER can be submitted again.

~~Making a draft as I'd like to add some tests and some cleanup is required, but can be reviewed regarding the overall direction.~~

TODO:

-  [x] Add some tests for caching the reasonable receipts(out-of-order issue). Some details are expanded in the commit message, makes more sense to also take a look.
   - https://github.com/subspace/subspace/pull/436/commits/38b462084ce7fface29fb2a761470d0bebd05a9c discovered while writing the test

liuchengxu committed:
Ensure the block at which the receipt was generated is part of the history

Basically, it means the header of block specied by the primary number
of the execution receipt exists on the primary chain. In order to check
this, we have to access the client function within the runtime.

The execution receipt signer check can also be implemented in this
extension later.

liuchengxu committed:
Ensure the receipt is neither old nor too new

While validating the unsigned extrinsic in the tx pool:

- If the primary number of a receipt is not larger than the best known
  execution chain number, it will be rejected as being stale.
- If the primary number of a receipt is larger than the best known
  execution chain number plus the maximum drift, it will be reject as
  being too far in the future.

Before dispatching the extrinsic `submit_execution_receipt`, check the
primary number of receipt is exactly `best_execution_chain_number + 1`.

liuchengxu committed:
Add `primary_number` to FraudProof

So that the receipts after the corrected one will be removed as being
invalid once the fraud proof is verified to be correct.

liuchengxu committed:
Revert the execution chain if fraud proof is valid

- Reset the execution chain best number.
- Remove the receipts that is newer than the new execution chain best
  number

liuchengxu committed:
Extract `validate_fraud_proof`

liuchengxu committed:
Reject the fraud proof if receipt has been pruned or from the future

vedhavyas reviewed (commented):
LGTM. any further changes to go in ?

liuchengxu committed:
Expose transaction pool in `NewFull` for the test purpose

liuchengxu committed:
Allow specifying the executor node role

The semantic changes during the previous series refactoring on the
executor, this patch allows to run the executor in different mode again
without significant changes. The long term goal is to upgrade the cirrus
test service accordingly, but I think it's not a priority and actually
not a job I'm interested in :D.

liuchengxu committed:
Fix the execution receipt age check

- The receipt must point to a previous block instead of the current one.
- We should compare the primary number of a future receipt againest the
  best execution chain number instead of the best consensus chain number.

liuchengxu committed:
Add test for rejecting the receipts that are too far in the future

This also unlocks more transaction pool tests.

nazar-pc reviewed (commented):
> Basically, it means the header of block specied by the primary number
of the execution receipt exists on the primary chain. In order to check
this, we have to access the client function within the runtime.

I'm not sure this is the case. We have `ReceiptsPruningDepth` already, we can use it to limit the size of the cache of parent blocks and avoid adding host functions. I think host functions should be last resort tool in case it is not possible to achieve something without it.

liuchengxu committed:
Check the existence of block ER points to with a much simpler way

Thanks for the link mentioned https://github.com/subspace/subspace/pull/436#issuecomment-1125628608,
I was unaware of this and used an unnecessarily much more complex approach for the same purpose.

nazar-pc commented on crates/pallet-executor/src/lib.rs:
```diff
@@ -142,14 +156,19 @@ mod pallet {
             // Remove the oldest once the receipts cache is full.
             if let Some(to_prune) = primary_number.checked_sub(&T::ReceiptsPruningDepth::get()) {
                 Receipts::<T>::remove(to_prune);
+                OldestReceiptNumber::<T>::put(to_prune + 1u32.into());
```
This seems to be more canonical in Substrate (ad similar code above):
```suggestion
                OldestReceiptNumber::<T>::put(to_prune + One::one());
```

nazar-pc commented on crates/pallet-executor/src/lib.rs:
```diff
@@ -192,7 +211,12 @@ mod pallet {
+            <ExecutionChainBestNumber<T>>::put(new_best);
+            Receipts::<T>::iter_keys()
+                .filter(|b| b > &new_best)
+                .for_each(Receipts::<T>::remove);
```
I think this can be more efficient if you iterate starting with `new_best + One::one()` and use `::take()` for removal until it returns `None`. Then you'll read less stuff (in best case just one key which doesn't exist).

Also description on `type Receipts` can be expanded describing that receipts are not stored forever (due to pruning).

liuchengxu replied:
Updated to use `mutate` so that both the beginning and end can be specified explicitly.

nazar-pc commented on crates/pallet-executor/src/lib.rs:
```diff
@@ -294,11 +327,18 @@ mod pallet {
     #[repr(u8)]
-    enum InvalidTransactionCode {
+    pub enum InvalidTransactionCode {
         BundleEquivicationProof = 101,
```
BTW, is there a reason to start with 101 instead of `0`?

liuchengxu replied:
The intention is to let the error code of each pallet have its own scope of errors, otherwise, we can't precisely tell the error with just the custom error code if multiple pallets define the error from `0`, some substrate pallets already use `0`.

nazar-pc replied:
Hm, if we just have `0..=255` available here, start at `101` feels like an arbitrary choice. In which context we might have this ambiguity of error codes, isn't it transaction-specific?

liuchengxu replied:
Yes, it's an arbitrary choice at present, we can redesign this to make it more formal when the codebase becomes more stable. All invalid transactions using custom error collapse to this global `Custom(u8)` type, it's not pallet or transaction-specfic.

nazar-pc commented on crates/pallet-executor/src/lib.rs:
```diff
@@ -192,7 +211,12 @@ mod pallet {
             );
 
-            // TODO: revert the execution chain.
+            // Revert the execution chain.
```
It seems to me that pre-dispatch and validation logic will prevent this revert from actually going through. Am I missing something?

liuchengxu replied:
Hmm, which part are you referring to? For `submit_fraud_proof`, pre-dispatch is a mere `Ok(())`, validate_unsigned only checks the proof validity and the receipt has no been pruned.

nazar-pc replied:
I'm sorry, you're right, I didn't expand a piece of code to see that this is a new call.

nazar-pc commented on crates/pallet-executor/src/lib.rs:
```diff
@@ -430,6 +511,21 @@ impl<T: Config> Pallet<T> {
+        }
+
+        let current_block_number = frame_system::Pallet::<T>::current_block_number();
+        if primary_number == current_block_number
+            || primary_number > best_number + T::MaximumReceiptDrift::get()
```
Shouldn't it just be this?:
```suggestion
        if primary_number > current_block_number
```

I think `current_block_number` is the previous block, so whenever we receive ER in the very next block `primary_number == current_block_number` will render ER invalid and cause it to be accepted with a delay in the next block after this.

And I do not understand why you need a drift here that might run past `current_block_number`, which in turn should be physically impossible and thus invalid.

liuchengxu replied:
Okay, a more descriptive name might be helpful here. `current_block_number` is the current consensus chain number, `best_number` is the best execution chain number. The execution chain can lag behind the consensus chain due to no ERs being submitted in time.

- `primary_number == current_block_number` is because the receipt must point to a previous block, you can't submit a receipt for a block that is being built right now.
- `primary_number > best_number + T::MaximumReceiptDrift::get()` means we only accept a reason number of future ERs that are ahead of the current best execution chain number. The new receipts should extend the execution chain instead of the consensus chain, so we compare it against the execution chain best number.

nazar-pc replied:
> current_block_number is the current consensus chain number

I do not think so. It is the tip of the consensus chain. Validation is called by transaction pool to validate transactions and I don't think it is called within block production from within runtime, rather transaction pools calls here to check whether transaction became valid after new block was produced.

> means we only accept a reason number of future ERs that are ahead of the current best execution chain number

But why? `best_number + T::MaximumReceiptDrift::get()` can easily be higher than `current_block_number`, which doesn't make sense.

The check I suggested seems to cover both of these cases in a simpler to understand way.

liuchengxu replied:
>I do not think so. It is the tip of the consensus chain. Validation is called by transaction pool to validate transactions and I don't think it is called within block production from within runtime, rather transaction pools calls here to check whether transaction became valid after new block was produced.

Yeah, I meat the tip of the chain, but said in a confusing way. Validation is not called within block production but it is called against the tip of the chain to see if they are ready to be included in the next block, for example, the context for performing the validation is the latest imported number + 1(the next block number) when you submit an extrinsic.

>But why? best_number + T::MaximumReceiptDrift::get() can easily be higher than current_block_number, which doesn't make sense.

There is a case that the local height is lower than the network height due to the network delay, but somehow it received the receipts for the valid futures blocks so we cache a number of these receipts in the transaction pool, no? Simply using `if primary_number > current_block_number` will just reject these kind of receipts unconditionally.

nazar-pc replied:
> Yeah, I meat the tip of the chain, but said in a confusing way. Validation is not called within block production but it is called against the tip of the chain to see if they are ready to be included in the next block, for example, the context for performing the validation is the latest imported number + 1(the next block number) when you submit an extrinsic.

I had to test it, but you're right, I had an impression that it'll test against current block number. `primary_number >= current_block_number` then?

> There is a case that the local height is lower than the network height due to the network delay, but somehow it received the receipts for the valid futures blocks so we cache a number of these receipts in the transaction pool, no? Simply using if primary_number > current_block_number will just reject these kind of receipts unconditionally.

From what I understand it will consider them invalid for now, but then transaction might become valid later, no? If so, then additional concept of `MaximumReceiptDrift` is unnecessary.

liuchengxu replied:
>primary_number >= current_block_number then?

Unless we clearly reject any potential valid future receipts(`MaximumReceiptDrift`), `primary_number >= current_block_number` will cause conflict with `primary_number > best_number + T::MaximumReceiptDrift::get()`.

>From what I understand it will consider them invalid for now, but then transaction might become valid later, no? If so, then additional concept of MaximumReceiptDrift is unnecessary.

It's not considered as invalid, but will put into the futures queue which requires some prerequisites to be satisfied, see the test case added in https://github.com/subspace/subspace/pull/436/commits/11dbaf85817f2c9ee780a12d0874c70e34e8012e. From what I understand, this is the out-of-order issue caused by the network delay as we discussed before, unless I have some misunderstanding here.

nazar-pc replied:
OK, I understand what you said now and satisfied with the answer, thanks!

nazar-pc commented on crates/pallet-executor/src/lib.rs:
```diff
@@ -455,6 +551,24 @@ impl<T: Config> Pallet<T> {
+        ensure!(
+            to_prove >= OldestReceiptNumber::<T>::get()
+                && to_prove <= ExecutionChainBestNumber::<T>::get(),
+            Error::<T>::ExecutionReceiptPrunedOrFuture
```
Out of curiosity, why separate error variants for ERs and one for FP?

liuchengxu replied:
Because I didn't find a satisfying naming for `ExecutionReceiptTooFarInFutureForFraudProof` 🙈  Any suggestion?

nazar-pc replied:
I meant why not `ExecutionReceiptPruned` and `ExecutionReceiptInFuture` separately :shrug:

liuchengxu replied:
`ExecutionReceiptTooFarInFuture` has already been used for `submit_execution_receipt`, using `ExecutionReceiptInFuture` for FP is not enough to distinguish these two and I'm not sure it's a good idea to use one kind for two cases even they share some similarity. https://github.com/subspace/subspace/blob/64afe53c02a546872a4ae20e10d2edbf76fe8e62/crates/pallet-executor/src/lib.rs#L540

nazar-pc replied:
I mean it could have been `FraudProofErInFuture` or something, but you chose to combine two cases into one anyway, so I was wondering why.
Another option is to have ER error variant and FP error variant and have another enum inside of those to have clear separation. But since events can be mapped to extrinsics, I'm not sure using the same variant is even an issue in the first place.

liuchengxu replied:
Thanks for the clue, the nested error looks much better!

liuchengxu committed:
Use One::one() and add some docs

liuchengxu committed:
Fix requirs tag for `submit_execution_receipt` and revamp the test

liuchengxu committed:
Optimize the invalid receipts removal

liuchengxu committed:
Make use of the nested pallet error in pallet-executor

liuchengxu committed:
Fix pallet-executor test
